File: PI_Words/PI/PiGenerator.py
#!/usr/bin/python3

import math

global DEBUG
DEBUG = False

if DEBUG:
    def debug(printstring):
        print(printstring)
else:
    def debug(doesntmatter):
        pass

class PiGenerator:
    ''' Returns precision hexadecimal digits of the fractional part of pi.
     Returns digits in most significant to least significant order.

     If precision < 0, return null.

     @param precision The number of digits after the decimal place to
                      retrieve.
     @return precision digits of pi in hexadecimal.
    '''

    @staticmethod
    def computePiInHex(precision):
        # TODO: Implement (Problem 1.d)
        if precision <= 0:
            return []
        PiInHex = []
        for i in range(precision):
            PiInHex.append(PiGenerator.piDigit(i))
        return PiInHex

    # decorator for python 2! unbound method removed from python 3!
    @staticmethod
    def powerMod(a, b, m):
        ''' Computes a^b mod m

        If a < 0, b < 0, or m < 0, return -1.

        @return a^b mod m
        TODO: Implement (Problem 1.b)'''

        if a | b | m < 0:
            return -1
        # power = math.floor(a**b)
        # carryover = power % m
        # carryover = power % m
        result = pow(a,b,m)
        debug("== return powerMod: {0} ==".format(result));
        return result

    @staticmethod
    def alternate_powerMod(x, y, z):
        '''Calculate (x ** y) % z efficiently.

        http://stackoverflow.com/questions/5246856/

        is 1 second slower for pi_precision = 1000 (8.4 seconds total)
        '''
        number = 1
        while y:
            if y & 1:
                number = number * x % z
            y >>= 1
            x = x * x % z
        return number

    @staticmethod
    def piDigit(n):
        '''
        Computes the nth digit of Pi in base-16.

        If n < 0, return -1.

        @param n The digit of Pi to retrieve in base-16.
        @return The nth digit of Pi in base-16.

        NOTE: testing done in ../test/test_PiGeneratorTest.py|35|
        '''

        if n < 0:
            return -1
        n -= 1
        x = 4 * PiGenerator.piTerm(1, n) - 2 * PiGenerator.piTerm(4, n) - \
                   PiGenerator.piTerm(5, n) - PiGenerator.piTerm(6, n)
        x = x - math.floor(x)
        result = math.floor(x * 16)
        debug("-- return piDigit: {0} --".format(result));
        return result

    @staticmethod
    def piTerm(j, n):
        # Calculate the left sum
        debug("++ piTerm j, n: {0}, {1}".format(j, n))
        s = 0
        for k in range(0, n+1):
            r = 8 * k + j
            s += PiGenerator.powerMod(16, n-k, r) / r
            s = s - math.floor(s)
        # Calculate the right sum
        t = 0
        k = n+1
        # Keep iterating until t converges (stops changing)
        while 1:
            r = 8 * k + j
            newt = t + math.pow(16, n-k) / r
            if (t == newt):
                break
            else:
                t = newt
            k += 1
        result = s+t
        debug("++ return piTerm: {0} ++".format(result));
        return result

File: java_sources/src/piwords/PiGenerator.java
package piwords;

import java.math.BigInteger;

public class PiGenerator {
    /* package piwords; */
    /**
     * Returns precision hexadecimal digits of the fractional part of pi.
     * Returns digits in most significant to least significant order.
     * 
     * If precision < 0, return null.
     * 
     * @param precision The number of digits after the decimal place to
     *                  retrieve.
     * @return precision digits of pi in hexadecimal.
     */
    public static int[] computePiInHex(int precision) {
        // TODO: Implement (Problem 1.d)
        //
        int[] PiInHex = new int[precision];
        for (int i = 0; i < precision; i++){
            PiInHex[i] = PiGenerator.piDigit(i);
        }
        return PiInHex;
    }

    /**
     * Computes a^b mod m
     * 
     * If a < 0, b < 0, or m < 0, return -1.
     * 
     * @param a
     * @param b
     * @param m
     * @return a^b mod m
     */
    /* public static int powerMod(int a, int b, int m) { */
    /*     // TODO: Implement (Problem 1.b) */
    /*     return -1; */
    /* } */

    /* From online compiler */
    /* http://www.compileonline.com/compile_java_online.php */

    /* public static void main(String []args){ */
    /*     System.out.println(); */

    /*     System.out.print(piDigit(0)); */
    /*     /* System.out.printf("%n NEW DIGIT %n"); */
    /*     System.out.print(" "); */
    /*     System.out.print(piDigit(1)); */
    /*     /* System.out.printf("%n NEW DIGIT %n"); */
    /*     System.out.print(" "); */
    /*     System.out.print(piDigit(2)); */
    /*     System.out.print(" "); */
    /*     System.out.print(piDigit(3)); */
    /*     System.out.print(" "); */
    /*     System.out.print(piDigit(4)); */
    /*     System.out.print(" "); */
    /*     System.out.println(); */
    /* } */

    public static double powerMod(int a, int b, int m) {
        /*''' Computes a^b mod m

          If a < 0, b < 0, or m < 0, return -1.

          @param a
          @param b
          @param m
          @return a^b mod m
    TODO: Implement (Problem 1.b)'''*/
        double carryover;
        double power;

        if (a < 0 | b < 0| m < 0) return -1;
        /* power = Math.pow(a, b); */
        /* carryover = power % m; */
        BigInteger a_big = BigInteger.valueOf(a);
        BigInteger b_big = BigInteger.valueOf(b);
        BigInteger m_big = BigInteger.valueOf(m);
        BigInteger result_big = a_big.modPow(b_big, m_big);
        int result = result_big.intValue();
        /* System.out.printf("== return powerMod: %f ==%n", carryover); */
        return result;
    }

    /**
     * Computes the nth digit of Pi in base-16.
     * 
     * If n < 0, return -1.
     * 
     * @param n The digit of Pi to retrieve in base-16.
     * @return The nth digit of Pi in base-16.
     */
    public static int piDigit(int n) {
        if (n < 0) return -1;

        n -= 1;
        double x = 4 * piTerm(1, n) - 2 * piTerm(4, n) -
            piTerm(5, n) - piTerm(6, n);
        x = x - Math.floor(x);
        int result = (int)(x * 16); 
        /* System.out.printf("-- return piDigit: %d --%n", result); */
        return result;
    }

    private static double piTerm(int j, int n) {
        // Calculate the left sum
        double s = 0;
        for (int k = 0; k <= n; ++k) {
            int r = 8 * k + j;
            s += powerMod(16, n-k, r) / (double) r;
            s = s - Math.floor(s);
        }

        // Calculate the right sum
        double t = 0;
        int k = n+1;
        // Keep iterating until t converges (stops changing)
        while (true) {
            int r = 8 * k + j;
            double newt = t + Math.pow(16, n-k) / r;
            if (t == newt) {
                break;
            } else {
                t = newt;
            }
            ++k;
        }
        double result = s+t;
        /* System.out.printf("++ return piTerm: %f ++%n", result); */
        return result;
    }
}
